src/Persistence/MsApp/MsappArchive.cs

```diff
@@ -452,24 +455,49 @@ public void Save(Control control, string? directory = null)
 
     private string GetSafeEntryPath(string directory, string name, string extension)
     {
-        var safeName = SafeFileNameRegex().Replace(name, "").Trim();
-        if (string.IsNullOrWhiteSpace(safeName))
+        if (!TryMakeSafeForEntryPathSegment(name, out var safeName, unsafeCharReplacementText: ""))
             throw new ArgumentException("Control name is not valid.", nameof(name));
 
-        var entryPath = Path.Combine(directory, $"{safeName}{extension}");
-        if (!CanonicalEntries.ContainsKey(NormalizePath(entryPath)))
-            return entryPath;
+        if (!TryGenerateUniqueEntryPath(directory.WhiteSpaceToNull(), safeName, extension, out var entryPath))
+        {
+            throw new InvalidOperationException("Failed to find a unique name for the control.");
+        }
+
+        return entryPath;
+    }
+
+    /// <inheritdoc/>
+    public bool TryGenerateUniqueEntryPath(
+        string? directory,
+        string fileNameNoExtension,
+        string? extension,
+        [NotNullWhen(true)] out string? entryPath,
+        string uniqueSuffixSeparator = "")
+    {
+        if (directory != null && string.IsNullOrWhiteSpace(directory))
+        {
+            throw new ArgumentException("The directory can be null, but cannot be empty or whitespace only.", nameof(directory));
+        }
+        _ = !string.IsNullOrEmpty(fileNameNoExtension) ? fileNameNoExtension : throw new ArgumentNullException(nameof(fileNameNoExtension));
+
+        var entryPathPrefix = directory == null ? fileNameNoExtension : Path.Combine(directory, fileNameNoExtension);
+
+        // First see if we can use the name as is
+        entryPath = $"{entryPathPrefix}{extension}";
+        if (!DoesEntryExist(entryPath))
```

Comment:
This checks raw path against canonical entries which can be an issue if path is "/entryA" for example